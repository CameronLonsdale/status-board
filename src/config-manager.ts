import * as bug from 'debug';
import * as fs from 'fs';
import * as path from 'path';
import * as extend from 'xtend';
import { logger } from './logger';

const debug = bug('config-manager');

function readConfigIfExists(fileName: string) {
  let configFileName = fileName;

  if (!path.extname(configFileName)) {
    configFileName = configFileName + '.js';
  }

  if (fs.existsSync(configFileName)) {
    return require(configFileName);
  }

  return {};
}

function readEnv(configFileName: string) {
  const key = 'ATLASBOARD_CONFIG_' + configFileName;

  debug('ENV key', key);

  const environmentKey = process.env[key];

  if (environmentKey) {

    debug('ENV configuration found for', key);

    try {
      const configValue = JSON.parse(environmentKey);

      if (typeof configValue === 'object') {
        return JSON.parse(environmentKey);
      }

      logger.error(`
        ENV configuration key ${key} could not be serialized into an object: ${environmentKey}
      `);

    } catch (error) {
      throw Error(`
        ENV configuration key ${key} contains invalid JSON: ${environmentKey}
        Error: ${error}
      `);
    }
  }
}

export default function (configFileName: any) {
  const localConfigFilePath = path.join(process.cwd(), 'config', configFileName);
  const atlasboardConfigFilePath = path.join(__dirname, '../config/', configFileName);

  return extend(
    readConfigIfExists(atlasboardConfigFilePath),
    readConfigIfExists(localConfigFilePath),
    readEnv(configFileName),
  );
}
