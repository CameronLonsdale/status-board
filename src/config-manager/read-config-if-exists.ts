import * as fs from 'fs';
import * as path from 'path';

export function readConfigIfExists(fileName: string) {
  let configFileName = fileName;

  if (!path.extname(configFileName)) {
    configFileName = configFileName + '.js';
  }

  if (fs.existsSync(configFileName)) {
    return require(configFileName);
  }

  return {};
}
