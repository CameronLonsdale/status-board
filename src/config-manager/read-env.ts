import * as bug from 'debug';

const debug = bug('config-manager');

export function readEnv(configFileName: string) {
  const key = 'ATLASBOARD_CONFIG_' + configFileName;

  debug('ENV key', key);

  const environmentKey = process.env[key];

  if (environmentKey) {

    debug('ENV configuration found for', key);

    try {
      return JSON.parse(environmentKey);
    } catch (error) {
      throw Error(`
        ENV configuration key ${key} contains invalid JSON: ${environmentKey}
        Error: ${error}
      `);
    }
  }
}
