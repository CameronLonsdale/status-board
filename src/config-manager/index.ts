import * as path from 'path';
import * as extend from 'xtend';
import { readConfigIfExists } from './read-config-if-exists';
import { readEnv } from './read-env';

export function configManager(configFileName: any) {
  const localConfigFilePath = path.join(process.cwd(), 'config', configFileName);
  const atlasboardConfigFilePath = path.join(__dirname, '../../config/', configFileName);

  return extend(
    readConfigIfExists(atlasboardConfigFilePath),
    readConfigIfExists(localConfigFilePath),
    readEnv(configFileName),
  );
}
