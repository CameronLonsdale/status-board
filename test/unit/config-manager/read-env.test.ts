import { readEnv } from '../../../src/config-manager/read-env';

describe('Config Manager: Read Env', () => {
  test('Should match snapshot', () => {
    expect(readEnv).toMatchSnapshot();
  });
});
