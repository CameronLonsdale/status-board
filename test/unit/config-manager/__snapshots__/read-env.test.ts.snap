// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Manager: Read Env Should match snapshot 1`] = `[Function]`;
